[28952] OAuth provider for API authentication

oliverguenther opened the pull request:
Basing on https://github.com/opf/openproject/pull/2408 , implements an OAuth adapter for Warden using Doorkeeper.

![screenshot from 2018-11-13 16-08-25](https://user-images.githubusercontent.com/459462/48422250-8dfe5e80-e75e-11e8-9676-33142710c3e6.png)
![screenshot from 2018-11-13 16-08-55](https://user-images.githubusercontent.com/459462/48422252-8f2f8b80-e75e-11e8-81e3-f0a17803f8f1.png)
![screenshot from 2018-11-13 16-09-16](https://user-images.githubusercontent.com/459462/48422254-8fc82200-e75e-11e8-88ba-f03f6d7b2341.png)
![screenshot from 2018-11-13 16-09-23](https://user-images.githubusercontent.com/459462/48422257-90f94f00-e75e-11e8-8f83-f90724f98372.png)


**TODOS**

- [x] OAuth applications and grants / revocation CRUD.
- [x] Adapt Doorkeeper as warden strategy to allow APIv3 access
- [x] Wire apiv3 scope to doorkeeper scope
- [x] Fix reversed view order breaking overridden doorkeeper views https://github.com/opf/openproject/pull/6824
- [x] Look into reusing endpoints for BCF. Can be solved by specifying OAuth urls in the BCF API endpoints
- [x] Complete tests with doorkeeper helpers
- [X] Decide on displaying and storage of client_secret after initial generation. Doorkeeper doesn't provide hashing of either tokens or client details, so we'd need to override model access in that case. PR opened at https://github.com/doorkeeper-gem/doorkeeper/pull/1168
- [x] Cleanup residual tokens after they have expired or were revoked. Cleanup happens in a delayed job after each successful authorization (= of a new application)

oliverguenther committed:
Add OAuth with warden

oliverguenther committed:
Implement OAuth applications CRUD

oliverguenther committed:
Allow engines to override core

oliverguenther committed:
Add rake task job that will dynamically load tasks

oliverguenther committed:
Show application secrets only once

oliverguenther committed:
Use the override core paths in the plugin

ulferts commented on config/locales/en.yml:
```diff
@@ -2580,3 +2601,84 @@ en:
+          The allowed URLs authorized users can be redirected to. One entry per line.
+          <br/>
+          If you're registering a desktop application, use the following URL.
+        confidential: "Check if Application will be used where the client secret can be kept confidential. Native mobile apps and Single Page Apps are assumed non-confidential."
```
s/Application/the application

ulferts commented on config/locales/en.yml:
```diff
@@ -2580,3 +2601,84 @@ en:
+          <br/>
+          If you're registering a desktop application, use the following URL.
+        confidential: "Check if Application will be used where the client secret can be kept confidential. Native mobile apps and Single Page Apps are assumed non-confidential."
+        scopes: Select which scopes this application can access. Leave empty to use defaults.
```
What are the defaults?

ulferts commented on config/locales/en.yml:
```diff
@@ -2580,3 +2601,84 @@ en:
+    client_id: "Client ID"
+    client_secret_notice: >
+      This is the only time we can print the client secret, please note it down and keep it secure.
+      It should be treated as a password and cannot be retrieved by OpenProject it at a later time.
```
s/by OpenProject it at/OpenProject at

ulferts commented on app/models/user.rb:
```diff
@@ -79,6 +79,16 @@ class User < Principal
+  # User-defined oauth applications
+  has_many :oauth_applications,
+           class_name: 'Doorkeeper::Application',
+           as: :owner
```
What will happen with those if the user is deleted?

oliverguenther replied:
Owner will be nullified. IMO that's okay, since it's only to track which admin created an application.

ulferts replied:
What do we need the owner for anyway? Displaying the user who created the app is not that valuable IMO. But I have no strong opinion there.

ulferts commented on config/initializers/menus.rb:
```diff
@@ -186,6 +190,11 @@
+  menu.push :oauth_applications,
+            { controller: '/oauth/applications', action: 'index' },
+            html: { class: 'oauth_applications' },
+            icon: 'icon2 icon-key'
```
It might be worthwhile (albeit probably out of scope) to group "LDAP authentication", "Two-factor authentication", "OAuth applications" and what is currently under "System settings -> Authentication" under a common menu item to help structuring the admin menu

oliverguenther replied:
The parent menu item would then also need to 'be' something, we currently can't group a set of menus without such a parent page

ulferts replied:
We could take most of what is in "System settings -> Authentication" to make up the content for the parent menu item. And while we are at it, we could move 
![image](https://user-images.githubusercontent.com/617519/51037770-705c4280-15b1-11e9-8992-f1bc0967ddd2.png)
out of the authentication settings as they are not related to it.

ulferts commented on config/routes.rb:
```diff
@@ -341,6 +347,7 @@
     end
     resources :enumerations
 
+
```
please remove

ulferts commented on config/routes.rb:
```diff
@@ -508,6 +519,11 @@
   # alternate routes for the current user
   scope 'my' do
     match '/deletion_info' => 'users#deletion_info', via: :get, as: 'delete_my_account_info'
+
```
please remove extra blank line

ulferts commented on spec/features/admin/oauth/oauth_user_flow_spec.rb:
```diff
@@ -0,0 +1,101 @@
+    # And also have no grant for this application
+    user.oauth_grants.reload
+    expect(user.oauth_grants.count).to eq 1
+    expect(user.oauth_grants.first.application).to eq app
```
specing a request done with the oauth credentials would be beneficial.

ulferts commented on config/initializers/doorkeeper.rb:
```diff
@@ -0,0 +1,219 @@
+  # https://github.com/doorkeeper-gem/doorkeeper/wiki/Using-Scopes
+  #
+  default_scopes  :public
+  optional_scopes :api_v3
```
we talked about removing the public scope. If that is done, the options could also be removed in the UI

ulferts commented on config/locales/en.yml:
```diff
@@ -2580,3 +2601,84 @@ en:
+    authorization_dialog:
+      authorize: "Authorize"
+      cancel: "Cancel and deny authorization."
+      prompt_html: "Authorize <strong>%{application_name}</strong> to use your account <em>%{login}</em> ?"
```
`s/</em> ?/</em>?`

ulferts commented on config/initializers/warden.rb:
```diff
@@ -26,5 +28,5 @@
 }
 OpenProject::Authentication.update_strategies(API_V3, api_v3_options) do |_strategies|
-  %i[global_basic_auth user_basic_auth basic_auth_failure session anonymous_fallback]
+  %i[global_basic_auth user_basic_auth basic_auth_failure oauth session anonymous_fallback]
```
I personally would have preferred it to be called `oauth2`

ulferts reviewed (changes requested):
Most of the functionality is looking good. Below are the bugs I noted:

#### Not displaying consent/redirecting to callback when not logged in
When not being logged in a request to the `authorize` endpoint will be redirected to the login screen. But after successful login, the browser is not being redirected to the consent screen or, if the consent has already been given, directly to the callback.
If the user has already been logged in, everything works fine.

#### bearer token generated by `client_credentials` not working 
The `client_credentials` flow creates a valid token but when using that token to access to API, the request is denied.
We already talked about linking the oauth app to a user account used for the `client_credentials` flow so this is more of a reminder to myself.

#### No consent required after revocation
When revoking an application, and after that issuing a request to the `authorize` endpoint, the user is not queried for their consent again.

#### No refresh token in `authorization_code` response
In response to a request to the `token` endpoint using the `authorization_code` flow, no `refresh_token` is issued. A response looks like this:
```
{
    "access_token": "575a002e990bee9087e907c8278ac73357aa511383d3aee6176b722aca8f4faa",
    "token_type": "Bearer",
    "expires_in": 7200,
    "scope": "api_v3",
    "created_at": 1547125162
}
```

#### Tokens usable after revocation
The access tokens issued before the user revokes their consent is still considered valid. This might actually conform to the specification as the token is a bearer token. But we should double check if new tokens can be generated using a refresh token once they are available (see before)

#### Missing margin when having consented to more than one app
![image](https://user-images.githubusercontent.com/617519/50971508-733d3180-14e4-11e9-9f59-168e28de5dd8.png)

#### Merge Access Token and OAuth Grants and Apps
![image](https://user-images.githubusercontent.com/617519/50971788-3291e800-14e5-11e9-8617-51af5af32ac8.png)

IMO, the menus for "Access tokens" and "OAuth Grants and Apps" should be merged as both are access tokens.

ulferts commented on config/locales/en.yml:
```diff
@@ -2580,3 +2601,84 @@ en:
+    authorization_error: "An authorization error has occurred."
+    my_application_grants: "Granted applications"
+    revoke_my_application_confirmation: "Do you really want to remove this application? This will revoke all %{count} tokens active for it."
+    my_oauth: "OAuth Grants and Apps"
```
s/OAuth Grants and Apps/OAuth grants and apps

oliverguenther committed:
Implement client credentials flow with a custom user_id

oliverguenther committed:
Enable refresh tokens

oliverguenther committed:
Revoke application in my page

oliverguenther commented:
I have removed individual grant revoking, I don't think it makes sense to do that. The access token page now only shows entire applications to revoke.

oliverguenther commented:
> Not displaying consent/redirecting to callback when not logged in

This happens by `ApplicationController` redirecting when login is required. I added a base controller to skip that action to properly use the session back_url for redirect.

> bearer token generated by client_credentials not working

A user_id can be assigned now. Ideally this would be an autocompleter, but I don't think we should go down that road now for such an edge case

> No consent required after revocation
> Tokens usable after revocation


Cannot reproduce after revoking entire application. Revoking individual grants is no longer possible

> No refresh token in authorization_code response

Because it was not enabled :slightly_smiling_face:  Do we want it? I assume so.

oliverguenther committed:
Fix foreign_key for MySQL

No idea why it wouldn't take `t.references` but this one works

oliverguenther committed:
Add client credentials spec

oliverguenther committed:
Fix specs after unifying

oliverguenther committed:
bump doorkeeper

oliverguenther committed:
Fix code flow spec to use integrated session

oliverguenther committed:
Add basic information on APIv3 authentication with OAuth2

ulferts commented on app/controllers/oauth/applications_controller.rb:
```diff
@@ -0,0 +1,116 @@
+    end
+
+    def find_app
+      @application = ::Doorkeeper::Application.find(params[:id])
```
This will lead to an unhandled `ActiveRecord::RecordNotFound` exception for non existing ids.

ulferts commented on docs/api/apiv3/introduction.apib:
```diff
@@ -59,6 +59,13 @@ API_KEY=2519132cdf62dcf5a66fd96394672079f9e9cad1
+- [Authorization code flow](https://oauth.net/2/grant-types/authorization-code)
+- [Client credentials](https://oauth.net/2/grant-types/client-credentials/) - Requires an application to be bound to an impersonating user for non-public access
+
 ### Why not username and password?
```
There is a "Future Work" section below which should be removed now.

ulferts commented on app/views/my/access_token.html.erb:
```diff
@@ -141,6 +141,31 @@ See docs/COPYRIGHT.rdoc for more details.
           <% end %>
+          <% granted_applications.each do |application, tokens| %>
+            <% latest_id = tokens.each_with_index.max(&:created_at)[1] %>
+            <% latest = tokens[latest_id] %>
```
This breaks (``NoMethodError: undefined method `created_at' for #<Array:0x00007f83d51956d0>``) once there are more than one token in tokens, e.g. when the user created a second token.

```
tokens.sort_by(&:created_at).last
```

should work

ulferts reviewed (changes requested):
> Not displaying consent/redirecting to callback when not logged in

> This happens by ApplicationController redirecting when login is required. I added a base controller to skip that action to properly use the session back_url for redirect.

Still does not work for me. If you are interested, we could take a look at it on my machine.

> No consent required after revocation
> Tokens usable after revocation

> Cannot reproduce after revoking entire application. Revoking individual grants is no longer possible

Working fine for me as well.

> No refresh token in authorization_code response

> Because it was not enabled 🙂 Do we want it? I assume so.

Thanks, working fine. :+1: And yes we definitely want it with the access tokens only being valid 2 hours.

> bearer token generated by client_credentials not working

Also working great.

> Tokens usable after revocation

This is still possible for access tokens but not for refresh tokens which is ok I guess but as we already have the access tokens stored in the db, why don't we check if they have been revoked?
https://www.oauth.com/oauth2-servers/listing-authorizations/revoking-access
https://www.oauth.com/oauth2-servers/access-tokens/self-encoded-access-tokens 

#### Scopes unnecessarily displayed in the UI

We currently only support the api scope. But the UI still displays the scope (as default scope) in the admin section:

![image](https://user-images.githubusercontent.com/617519/51038233-e0b79380-15b2-11e9-82bd-052a0e50d797.png)

![image](https://user-images.githubusercontent.com/617519/51038258-fa58db00-15b2-11e9-8bb9-632e815178ea.png)

#### User selection for client_credentials

Any chance we can get a proper select field there instead of forcing the user to have to retrieve the desired ID (and understanding first that he/she has to insert an id there) from somewhere else?

oliverguenther committed:
Review feedback

oliverguenther commented:
> This is still possible for access tokens but not for refresh tokens which is ok I guess but as we already have the access tokens stored in the db, why don't we check if they have been revoked?
> https://www.oauth.com/oauth2-servers/listing-authorizations/revoking-access
> https://www.oauth.com/oauth2-servers/access-tokens/self-encoded-access-tokens

We should raise this issue with doorkeeper then, it has nothing to do with this PR since we don't handle token lifetime.

oliverguenther commented:
> Any chance we can get a proper select field there instead of forcing the user to have to retrieve the desired ID (and understanding first that he/she has to insert an id there) from somewhere else

We currently only have the ng1 autocomplete-decoration which renders the entire users to the frontend. At least its not breaking that fast like a select over all users would, but it's not something I want to add.

Let's add an autocompleter on `/api/v3/users` once https://github.com/opf/openproject/pull/6988 is merged. [I created an appropriate ticket](https://community.openproject.com/wp/29404)

ulferts commented:
> Let's add an autocompleter on `/api/v3/users` once #6988 is merged. [I created an appropriate ticket](https://community.openproject.com/wp/29404)

:+1:

ulferts commented:
> We should raise this issue with doorkeeper then, it has nothing to do with this PR since we don't handle token lifetime.

I think that it would only require us to add a check like `!@token.revoked?` to https://github.com/opf/openproject/pull/6824/files#diff-ff74f90b1837049773886e5de58b4d1bR14

oliverguenther commented:
> I think that it would only require us to add a check like !@token.revoked? to https://github.com/opf/openproject/pull/6824/files#diff-ff74f90b1837049773886e5de58b4d1bR14

`token.accessible?` is already defined as `!expired? && !revoked?` https://github.com/doorkeeper-gem/doorkeeper/blob/master/lib/doorkeeper/models/concerns/accessible.rb#L11

ulferts committed:
ensure redirecting to oauth2 consent after login

Placing the back_url in the session does not suffice as session initialization in most cases will reset the back_url stored in the session. But even in those cases is the session value then set to the value in params

ulferts commented:
> Not displaying consent/redirecting to callback when not logged in

Fixed (at least on my machine) by 90cade5
